crates/prompt_gateway/src/http_context.rs

```diff
@@ -238,18 +243,69 @@ impl HttpContext for StreamContext {
     }
 
     fn on_http_response_body(&mut self, body_size: usize, end_of_stream: bool) -> Action {
-        trace!(
-            "recv [S={}] bytes={} end_stream={}",
-            self.context_id,
-            body_size,
-            end_of_stream
+        debug!(
+            "on_http_response_body: recv [S={}] bytes={} end_stream={}",
+            self.context_id, body_size, end_of_stream
         );
 
         if !self.is_chat_completions_request {
             debug!("non-gpt request");
             return Action::Continue;
         }
 
+        if self.time_to_first_token.is_none() {
+            self.time_to_first_token = Some(
+                SystemTime::now()
+                    .duration_since(UNIX_EPOCH)
+                    .unwrap()
+                    .as_nanos(),
+            );
+        }
+
+        if end_of_stream {
+            if let Some(traceparent) = self.traceparent.as_ref() {
+                let since_the_epoch_ns = SystemTime::now()
+                    .duration_since(UNIX_EPOCH)
+                    .unwrap()
+                    .as_nanos();
+
+                let traceparent_tokens = traceparent.split("-").collect::<Vec<&str>>();
+                if traceparent_tokens.len() != 4 {
+                    warn!("traceparent header is invalid: {}", traceparent);
+                    return Action::Continue;
+                }
+                let parent_trace_id = traceparent_tokens[1];
+                let parent_span_id = traceparent_tokens[2];
+                let mut trace_data = common::tracing::TraceData::new();
+                trace_data.add_span(Span {
+                    trace_id: parent_trace_id.to_string(),
+                    parent_span_id: Some(parent_span_id.to_string()),
+                    span_id: format!("{}", get_random_span_id()),
+                    name: "total_time".to_string(),
+                    start_time_unix_nano: format!("{}", self.start_upstream_llm_request_time),
+                    end_time_unix_nano: format!("{}", since_the_epoch_ns),
+                    kind: 1,
+                    attributes: vec![],
+                });
+
+                trace_data.add_span(Span {
```

Comment:
I think this should be an even in the total_time span. We shouldn't represent TTFT as a separate span because there is technically no new service work. https://opentelemetry.io/docs/concepts/signals/traces/#span-events